packages/server-wallet/src/onchain-service/store.ts

```diff
@@ -0,0 +1,183 @@
+    // mining, so look by nonce
+    const idx = existing.findIndex(
+      e => (e.data as providers.TransactionResponse | NoncedMinimalTransaction).nonce === tx.nonce
+    );
+    return Promise.resolve(
+      this._updateTransactions(channelId, idx, {status: 'failed', data: {...tx, reason}})
+    );
+  }
+
+  private _updateTransactions<T extends TransactionStatus>(
+    channelId: Bytes32,
+    previousRecordIndex: number,
+    data: TransactionModel<T>
+  ): void {
+    // Get existing channel transactions
+    const existing = this.transactions.get(channelId) || [];
+
+    // Update record
+    if (previousRecordIndex === -1) {
+      // No record of pending transaction found
+      existing.push(data);
+    } else {
+      // Update existing record
+      existing[previousRecordIndex] = data;
+    }
+    this.transactions.set(channelId, existing);
+  }
+}
+
+export class OnchainServiceStore implements OnchainServiceStoreInterface {
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  private readonly events: Map<string, ChannelEventRecord[]> = new Map();
+
+  // Stores transactions that are going to be sent to mempool
+  private readonly requestedTransactions: Map<
+    string,
+    MinimalTransaction & {nonce: number}
+  > = new Map();
+
+  // Stores transactions that are currently in the mempool
+  private readonly pendingTransactions: Map<string, providers.TransactionResponse> = new Map();
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  constructor() {}
+
+  public getEvents(channelId: Bytes32): Promise<ChannelEventRecord[]> {
+    return Promise.resolve(this.events.get(channelId) || []);
+  }
+
+  public getLatestEvent<T extends ContractEventName>(
+    channelId: Bytes32,
+    event: T
+  ): ChannelEventRecordMap[T] {
+    if (!this.hasChannel(channelId)) {
+      throw new Error('Channel not found');
+    }
+    const unsorted = (this.events.get(channelId) || []).filter(e => e.type === event);
+    const [latest] = unsorted.sort((a, b) => {
+      // TODO: Can I safely sort by block number here?
+      // TODO: Is there another generic way (ie timestamp) to return the latest
+      // without the store needing to know details about the event fields
+      if (a.type !== b.type) {
+        throw new Error('This should never happen');
+      }
+      switch (b.type) {
+        case 'Deposited': {
+          return b.destinationHoldings.sub(a.destinationHoldings).toNumber();
+        }
+        default: {
+          const e: never = b.type;
+          throw new Error(`Unrecognized contract event type: ${e}`);
+        }
+      }
+    });
+
+    return latest;
+  }
+
+  public saveEvent<T extends ContractEventName>(
+    channelId: Bytes32,
+    data: ChannelEventRecordMap[T]
+  ): Promise<void> {
+    const existing = this.events.get(channelId) || [];
+    const idx = existing.findIndex(e => {
+      return JSON.stringify(e) === JSON.stringify(data);
```

Comment:
Do events not have any other identifiers?

Reply:
Atm no, but @andrewgordstewart and I talked about restructuring the events and that was one of the things that came up. We're going to try to use the transaction hash.

One question I have is if transactions can be bundled and emit multiple events, since then we may want to use uuid as the identifier. Kind of took a shortcut here because I knew this would change pretty completely when its v1 with permanent storage